controllers/machine_controller_noderef.go

```diff
@@ -63,26 +59,65 @@ func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clust
 	if err != nil {
 		if err == ErrNodeNotFound {
+			// While a NodeRef is set in the status, getting ErrNodeNotFound error means the node is deleted.
+			// If Status.NodeRef is not set before, node still can be in the provisioning state.
+			if machine.Status.NodeRef != nil {
+				conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeNotFoundReason, clusterv1.ConditionSeverityError, "")
+				return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 20 * time.Second},
+					"failed to find the Node owned by Machine %q in namespace %q", machine.Name, machine.Namespace)
+			}
+
+			logger.Error(err, "Failed to assign NodeRef")
+			r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
 			return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 20 * time.Second},
 				"cannot assign NodeRef to Machine %q in namespace %q, no matching Node", machine.Name, machine.Namespace)
 		}
-		logger.Error(err, "Failed to assign NodeRef")
-		r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
+		// This may be a transient client error and may not mean there is no matching Node.
 		return err
 	}
 
 	// Set the Machine NodeRef.
-	machine.Status.NodeRef = nodeRef
-	logger.Info("Set Machine's NodeRef", "noderef", machine.Status.NodeRef.Name)
-	r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+	if machine.Status.NodeRef == nil {
+		machine.Status.NodeRef = &apicorev1.ObjectReference{
+			Kind:       node.Kind,
+			APIVersion: node.APIVersion,
+			Name:       node.Name,
+			UID:        node.UID,
+		}
+		logger.Info("Set Machine's NodeRef", "noderef", machine.Status.NodeRef.Name)
+		r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeProvisioningReason, clusterv1.ConditionSeverityWarning, "")
+		return nil
+	}
+
+	// Do the remaining node health checks, then set the node health to true if all checks pass.
+	if checkNodeConditions(node) == false {
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeConditionsFailedReason, clusterv1.ConditionSeverityWarning, "")
+		return nil
+	}
+	conditions.MarkTrue(machine, clusterv1.MachineNodeHealthyCondition)
+
 	return nil
 }
 
-func (r *MachineReconciler) getNodeReference(c client.Reader, providerID *noderefutil.ProviderID) (*apicorev1.ObjectReference, error) {
+// Return true if all Node Conditions are false other than "Ready" condition.
+// "Ready" being false does not mean there is a problem with the Node, may be missing CNI.
+func checkNodeConditions(node *apicorev1.Node) bool {
+	for _, condition := range node.Status.Conditions {
+		if condition.Type != apicorev1.NodeReady {
+			if condition.Status == apicorev1.ConditionTrue {
+				return false
+			}
+		}
+	}
```

Comment:
Should we check specific conditions, instead of ignoring `Ready` and checking all the others?

Reply:
Also we might want to add a bit more checks for control plane nodes?

Reply:
> Also we might want to add a bit more checks for control plane nodes?



I will have another PR that adds control-plane node conditions to the machines in `KubeadmControlPlane` controller. Since the owner of control planes machines are KCP, I think that will be a better place to add control-plane specific conditions there.

Reply:
+1 to check a list of conditions. I will include ready thought.

Reply:
I added the ready condition check too, but do we consider a Node unhealthy if it is not ready due to missing CNI?

Reply:
Probably not, although we don't have a way to detect if CNI has been installed or not. Checking `Ready` might also be overstepping our responsibilities with MachineHealthCheck, which already has timeouts for node ready

Reply:
My concern is the user perspective, and what I want to avoid is having the users being confused by the machine reporting the node is healthy, while the node itself reports being not ready
However, I plan to test this as soon as draft is lifted from the PR.

Reply:
Let's tackle this in a separated PR